### PR TITLE
Shape inference

### PR DESCRIPTION
**Pull request by castigli**
Adding missing parts for shape propagation. The goal is to setup a 3D serial transform.

### CODE REVIEW

File: ci/docker/README.md [hunk 1]
```diff
@@ -1,3 +1,6 @@
 # Build image with quiccir dependencies
+
+```bash
 docker build -f ci/docker/Dockerfile -t giacasti/ubuntu-llvm:22.04-17 .
-docker push giacasti/ubuntu-llvm:22.04-17
```
Comment (phmarti) [inline ci/docker/README.md:3]: We should not forget to put this into quicc docker repo when more or less stable

File: include/Quiccir/Interfaces/FoldTensorCastIntoConsumerOpInterface.td [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+//   Copyright (c) 2024,
+//   Earth and Planetary Magnetism group, ETH Zurich
+//
+//===---------------------------------------------------------------------===//
+//
+// Defines the operations of the operations that can fold a tensor producer.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef QUICCIR_INTERFACES_FOLDTENSORCASTINTOCONSUMEROPINTERFACE
```
Comment (phmarti) [inline include/Quiccir/Interfaces/FoldTensorCastIntoConsumerOpInterface.td:12]: Do you remove the file extension from the header guard on purpose?

File: include/Quiccir/Interfaces/ShapeInferenceOpInterface.td [hunk 3]
```diff
@@ -0,0 +1,32 @@
+//===- ShapeInferenceOpInterface.td - Shape Inference Interface -*- tablegen -==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
```
Comment (phmarti) [inline include/Quiccir/Interfaces/ShapeInferenceOpInterface.td]: Why are those files part of LLVM?

File: lib/Quiccir/Transforms/SetViewLayout.cpp [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+      // if (failed(deallocateBuffers(op)))
+      //   return WalkResult::interrupt();
+      return WalkResult::advance();
+    });
+
+  if (result.wasInterrupted())
+    signalPassFailure();
+
+
+  // if (failed(applyFullConversion(module, target, std::move(patterns))))
```
Comment (phmarti) [inline lib/Quiccir/Transforms/SetViewLayout.cpp]: Left over?

File: lib/Quiccir/Transforms/SetViewLayout.cpp [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+  //     signalPassFailure();
+}
+
+// std::unique_ptr<Pass> mlir::quiccir::createSetViewLayoutPass() {
```
Comment (phmarti) [inline lib/Quiccir/Transforms/SetViewLayout.cpp]: Left over?

File: lib/Quiccir/Transforms/SetViewLayout.cpp [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        setMissingLayout(mods, layout[2][1]);
+      }
+      if (auto jwPrjOp = dyn_cast<JWPOp>(op)) {
+        // set attributes if not set
+        Value phys = jwPrjOp.getPhys();
+        setMissingLayout(phys, layout[2][0]);
+        Value mods = jwPrjOp.getMods();
+        setMissingLayout(mods, layout[2][1]);
+      }
+      // if (failed(deallocateBuffers(op)))
```
Comment (phmarti) [inline lib/Quiccir/Transforms/SetViewLayout.cpp]: Left over?

**Review by phmarti — APPROVED** [review]
Are the .td files clang-formatted?

File: ci/docker/README.md (hunk 1 above)
Comment (castigli) [reply ci/docker/README.md:3]: added the image to the `quicc` docker repo

File: include/Quiccir/Interfaces/FoldTensorCastIntoConsumerOpInterface.td (hunk 2 above)
Comment (castigli) [reply include/Quiccir/Interfaces/FoldTensorCastIntoConsumerOpInterface.td:12]: I just followed the llvm convention, guards in `*.td` files have no extensions in the name

File: include/Quiccir/Interfaces/ShapeInferenceOpInterface.td (hunk 3 above)
Comment (castigli) [reply include/Quiccir/Interfaces/ShapeInferenceOpInterface.td]: fixed

File: lib/Quiccir/Transforms/SetViewLayout.cpp (hunk 4 above)
Comment (castigli) [reply lib/Quiccir/Transforms/SetViewLayout.cpp]: cleaned up

File: lib/Quiccir/Transforms/SetViewLayout.cpp (hunk 5 above)
Comment (castigli) [reply lib/Quiccir/Transforms/SetViewLayout.cpp]: cleaned up

**Comment by castigli** [comment]
> Are the .td files clang-formatted?

nothing is formatted, I need to add a file style, I think they added one upstream that I can copy. I'll add it as a separate PR.

File: lib/Quiccir/Transforms/SetViewLayout.cpp (hunk 6 above)
Comment (castigli) [reply lib/Quiccir/Transforms/SetViewLayout.cpp]: cleaned up

**Pull request merged**